src/main/java/com/wavefront/SpanGenerator.java

```diff
@@ -94,28 +94,97 @@ public SpanQueue generate(GeneratorConfig config) {
     int upperLevelSize;
     for (int n = levels - 1; n > 0; n--) {
       upperLevelSize = trace.get(n - 1).size();
-      for (Span childSpan :
-          trace.get(n)) {
-        childSpan.addParent(trace.get(n - 1).get(randGenerator.nextInt(upperLevelSize)));
+      for (Span childSpan : trace.get(n)) {
+        childSpan.addParent(trace.get(n - 1).get(RANDOM.nextInt(upperLevelSize)));
       }
     }
     return trace;
   }
 
-  protected List<Pair<String, String>> getTags(int errorRate) {
+  private List<Pair<String, String>> getTags(int errorRate) {
     // TODO: now it looks static, but for RCA we will generate tags variation for spans
-
     List<Pair<String, String>> tags = new LinkedList<>();
     tags.add(new Pair<>("application", "trace loader"));
     tags.add(new Pair<>("service", "generator"));
     tags.add(new Pair<>("host", "ip-10.20.30.40"));
 
     if (errorRate > 0) {
-      if (randGenerator.nextInt(100) < errorRate) {
+      if (RANDOM.nextInt(100) < errorRate) {
         tags.add(new Pair<>("error", "true"));
       }
     }
 
     return tags;
   }
+
+  /**
+   * Normalize all distributions. ie trace type, spans counts and so on.
+   */
+  private void normalizeDistributions(List<TraceTypePattern> traceTypePatterns) {
+    // trace types distribution
+    double sum = traceTypePatterns.stream().mapToDouble(d -> d.tracePercentage).sum();
+    traceTypePatterns.forEach(d ->
+        d.tracePercentage = (int) Math.round(d.tracePercentage * getNormalizationRatio(sum)));
```

Comment:
Why you don't calculate the Normalization Ratio at the beginning of the loop and avoid method call on each iteration?